internal/proxy/task_scheduler.go

```diff
@@ -228,16 +228,22 @@ type dmTaskQueue struct {
 func (queue *dmTaskQueue) Enqueue(t task) error {
 	queue.statsLock.Lock()
 	defer queue.statsLock.Unlock()
-	err := queue.addPChanStats(t)
+	//1. preAdd will check whether provided task is valid or addable
+	//and get the current pChannels for this dmTask
+	pChannels, dmt, err := queue.preAddPChanStats(t)
 	if err != nil {
 		return err
 	}
+	//2. enqueue dml task
 	err = queue.baseTaskQueue.Enqueue(t)
 	if err != nil {
-		queue.popPChanStats(t)
 		return err
 	}
-
+	//3. if preAdd succeed, commit will use pChannels got previously when preAdding and will definitely succeed
+	queue.commitPChanStats(dmt, pChannels)
```

Comment:
Will there be a problem with entering the queue first and then updating pChanStats?

Reply:
if ensue successfully but addPchanStats fail, there will an inconsistency